trin-core/src/portalnet/utp.rs

```diff
@@ -0,0 +1,1022 @@
+pub struct UtpListener {
+    pub discovery: Arc<Discovery>,
+    pub utp_connections: HashMap<ConnectionKey, UtpStream>,
+}
+
+impl UtpListener {
+    pub async fn process_utp_request(&mut self, payload: &[u8], node_id: &NodeId) {
+        match Packet::try_from(payload) {
+            Ok(packet) => {
+                let connection_id = packet.connection_id();
+
+                match packet.type_() {
+                    Type::StReset => {
+                        let key_fn = |offset| {
+                            ConnectionKey::new(node_id.clone(), connection_id - 1 + offset)
+                        };
+                        let f =
+                            |conn: &&mut UtpStream| -> bool { conn.conn_id_send == connection_id };
+
+                        if let Some(conn) = self.utp_connections.get_mut(&key_fn(1)) {
+                            conn.state = ConnectionState::Disconnected;
+                        } else if let Some(conn) =
+                            self.utp_connections.get_mut(&key_fn(2)).filter(f)
+                        {
+                            conn.state = ConnectionState::Disconnected;
+                        } else if let Some(conn) =
+                            self.utp_connections.get_mut(&key_fn(0)).filter(f)
+                        {
+                            conn.state = ConnectionState::Disconnected;
+                        }
+                    }
+                    Type::StSyn => {
+                        if let Some(enr) = self.discovery.discv5.find_enr(&node_id) {
+                            // If neither of those cases happened handle this is a new request
+                            let mut conn = UtpStream::init(self.discovery.clone(), enr);
+                            conn.handle_packet(packet).await;
+                            self.utp_connections.insert(
+                                ConnectionKey {
+                                    node_id: node_id.clone(),
+                                    conn_id_recv: conn.conn_id_recv,
+                                },
+                                conn,
+                            );
+                        } else {
+                            debug!("Query requested an unknown ENR");
+                        }
+                    }
+                    _ => {
+                        if let Some(conn) = self.utp_connections.get_mut(&ConnectionKey {
+                            node_id: node_id.clone(),
+                            conn_id_recv: connection_id,
+                        }) {
+                            conn.handle_packet(packet).await;
+                        }
+                    }
+                }
+            }
+            Err(e) => {
+                debug!("Failed to decode packet: {}", e);
+            }
+        }
+    }
+
+    // I am honestly not sure if I should init this with Enr or NodeId since we could use both
```

Comment:
If there's any reason you picked node_id in the end, some update would be nice. Otherwise, this comment is fine as is. (The comment is a little chatty, but kind of funny and actually conveys something important, if someone later decides that ENR might be better).

Reply:
I ended up picking node_id because it was cleaner and because they are more likely to have their node_id from what I understand